File: mcp-code-intelligence-nodejs/src/http/api-routes.ts
/**
 * REST API route handlers for memory engine — search, list, graph, stats.
 * Port of Kotlin MemoryApiRoutes.kt.
 */

import * as http from 'http';
import { MemoryEngine } from '../memory/memory-engine.js';
import { KnowledgeGraph } from '../memory/knowledge-graph.js';
import { KnowledgeEntry } from '../memory/models.js';

/** Dispatch API requests to the correct handler. */
export function handleApiRoute(
  url: URL,
  res: http.ServerResponse,
  engine: MemoryEngine | null,
  graph: KnowledgeGraph | null
): void {
  const path = url.pathname.replace('/api/memory', '');

  if (path === '/status') handleStatus(res, engine);
  else if (path === '/search') handleSearch(url, res, engine);
  else if (path.match(/^\/entries\/\d+$/)) handleGetEntry(path, res, engine);
  else if (path === '/entries') handleListEntries(url, res, engine);
  else if (path === '/sessions') handleSessions(url, res, engine);
  else if (path.match(/^\/sessions\/[^/]+\/events$/)) handleSessionEvents(path, url, res, engine);
  else if (path.match(/^\/graph\/\d+\/neighbors$/)) handleNeighbors(path, res, engine, graph);
  else if (path === '/graph/data') handleGraphData(url, res, engine, graph);
  else sendError(res, 404, 'Route not found');
}

function handleStatus(res: http.ServerResponse, engine: MemoryEngine | null): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const stats = engine.getStats();
  const tierMap: Record<string, number> = {};
  for (const t of stats.tierBreakdown) tierMap[t.tier] = t.entryCount;
  sendJson(res, {
    totalEntries: stats.totalEntries,
    totalEdges: stats.totalEdges,
    totalVectors: stats.totalVectors,
    tierBreakdown: tierMap,
  });
}

function handleSearch(url: URL, res: http.ServerResponse, engine: MemoryEngine | null): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const query = url.searchParams.get('q') ?? '';
  const limit = parseInt(url.searchParams.get('limit') ?? '10', 10);
  const tier = url.searchParams.get('tier');
  const results = tier
    ? engine.search.searchInTier(query, tier, limit)
    : engine.search.search(query, limit);
  sendJson(res, results.map(r => toEntryResponse(r.entry, r.score)));
}

function handleListEntries(url: URL, res: http.ServerResponse, engine: MemoryEngine | null): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const tier = url.searchParams.get('tier') ?? undefined;
  const type = url.searchParams.get('type') ?? undefined;
  const limit = parseInt(url.searchParams.get('limit') ?? '20', 10);
  const offset = parseInt(url.searchParams.get('offset') ?? '0', 10);
  const sort = url.searchParams.get('sort') ?? 'created_at';
  const afterId = url.searchParams.get('after_id');
  const entries = engine.knowledge.findFiltered(tier, type, limit, offset, sort, afterId ? parseInt(afterId, 10) : undefined);
  sendJson(res, entries.map(e => toEntryResponse(e)));
}

function handleGetEntry(path: string, res: http.ServerResponse, engine: MemoryEngine | null): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const id = parseInt(path.split('/')[2], 10);
  if (isNaN(id)) { sendError(res, 400, 'Invalid id'); return; }
  const entry = engine.knowledge.findById(id);
  if (!entry) { sendError(res, 404, 'Not found'); return; }
  engine.knowledge.recordAccess(id);
  sendJson(res, toDetailResponse(entry));
}

function handleNeighbors(
  path: string, res: http.ServerResponse,
  engine: MemoryEngine | null, graph: KnowledgeGraph | null
): void {
  if (!engine || !graph) { sendError(res, 503, 'Not initialized'); return; }
  const id = parseInt(path.split('/')[2], 10);
  if (isNaN(id)) { sendError(res, 400, 'Invalid id'); return; }
  const neighborIds = graph.getConnected(id);
  const entries = [...neighborIds]
    .map(nid => engine.knowledge.findById(nid))
    .filter((e): e is KnowledgeEntry => e !== undefined);
  sendJson(res, entries.map(e => toEntryResponse(e)));
}

function handleGraphData(
  url: URL, res: http.ServerResponse,
  engine: MemoryEngine | null, graph: KnowledgeGraph | null
): void {
  if (!engine || !graph) { sendError(res, 503, 'Not initialized'); return; }
  const stats = engine.getStats();
  if (stats.totalEdges === 0) {
    const entries = engine.knowledge.findByTier('WORKING', 50);
    const nodes = entries.map(e => toGraphNode(e));
    sendJson(res, { nodes, edges: [] });
    return;
  }
  const limit = parseInt(url.searchParams.get('limit') ?? '100', 10);
  const edges = engine.graphRepo.findAll(limit);
  const nodeIds = [...new Set(edges.flatMap(e => [e.source_id, e.target_id]))];
  const nodes = nodeIds
    .map(id => engine.knowledge.findById(id))
    .filter((e): e is KnowledgeEntry => e !== undefined)
    .map(e => toGraphNode(e));
  const edgeList = edges.map(e => ({ source: e.source_id, target: e.target_id, relation: e.relation }));
  sendJson(res, { nodes, edges: edgeList });
}

function handleSessions(url: URL, res: http.ServerResponse, engine: MemoryEngine | null): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const agent = url.searchParams.get('agent') ?? '';
  const status = url.searchParams.get('status') ?? '';
  const limit = parseInt(url.searchParams.get('limit') ?? '50', 10);
  const sessions = engine.sessions.listFiltered(agent, status, limit);
  sendJson(res, sessions.map(s => ({
    id: s.session_id, agentName: s.agent_name,
    startedAt: s.started_at, endedAt: s.ended_at,
    observationCount: s.observation_count, status: s.status,
  })));
}

function handleSessionEvents(
  path: string, url: URL, res: http.ServerResponse, engine: MemoryEngine | null
): void {
  if (!engine) { sendError(res, 503, 'Memory not initialized'); return; }
  const sessionId = path.split('/')[2];
  const limit = parseInt(url.searchParams.get('limit') ?? '200', 10);
  const events = engine.audit.listBySession(sessionId, limit);
  sendJson(res, events.map(e => ({
    id: e.id, operation: e.operation,
    entryId: e.entry_id, sessionId: e.session_id,
    details: e.details, createdAt: e.created_at,
  })));
}

/** Map entry to API response (without content). */
function toEntryResponse(e: KnowledgeEntry, score = 0) {
  return {
    id: e.id, summary: e.summary, type: e.type, tier: e.tier,
    confidence: e.confidence, accessCount: e.access_count,
    source: e.source, tags: e.tags, score,
  };
}

/** Map entry to detail response (with content). */
function toDetailResponse(e: KnowledgeEntry) {
  return {
    id: e.id, summary: e.summary, content: e.content,
    type: e.type, tier: e.tier, confidence: e.confidence,
    accessCount: e.access_count, source: e.source, tags: e.tags,
  };
}

/** Map entry to graph node response. */
function toGraphNode(e: KnowledgeEntry) {
  return { id: e.id, summary: e.summary.substring(0, 60), type: e.type, tier: e.tier, source: e.source };
}

function sendJson(res: http.ServerResponse, data: unknown): void {
  res.writeHead(200, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify(data));
}

function sendError(res: http.ServerResponse, code: number, message: string): void {
  res.writeHead(code, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: message }));
}

File: mcp-code-intelligence-python/src/mcp_code_intel/http/api_routes.py
"""REST API route handlers for memory engine — search, list, graph, stats."""

import re
from http.server import BaseHTTPRequestHandler

from ..memory.engine import MemoryEngine
from ..memory.knowledge_graph import KnowledgeGraph
from .response_helpers import (
    to_entry_response, to_detail_response, to_graph_node,
    send_json, send_error, first_param,
)
from .session_routes import handle_sessions, handle_session_events


def handle_api_route(
    path: str, query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
    graph: KnowledgeGraph | None,
) -> None:
    """Dispatch API requests to the correct handler."""
    try:
        _dispatch_api(path, query, handler, engine, graph)
    except Exception as e:
        import sys
        print(f"[http-api] Error: {e}", file=sys.stderr, flush=True)
        send_error(handler, 500, str(e))


def _dispatch_api(
    path: str, query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
    graph: KnowledgeGraph | None,
) -> None:
    """Internal dispatch — exceptions caught by handle_api_route."""
    api_path = path.replace("/api/memory", "")

    if api_path == "/status":
        _handle_status(handler, engine)
    elif api_path == "/search":
        _handle_search(query, handler, engine)
    elif re.match(r"^/entries/\d+$", api_path):
        _handle_get_entry(api_path, handler, engine)
    elif api_path == "/entries":
        _handle_list_entries(query, handler, engine)
    elif api_path == "/sessions":
        handle_sessions(query, handler, engine)
    elif re.match(r"^/sessions/[^/]+/events$", api_path):
        handle_session_events(api_path, query, handler, engine)
    elif api_path == "/audit":
        _handle_audit(query, handler, engine)
    elif re.match(r"^/graph/\d+/neighbors$", api_path):
        _handle_neighbors(api_path, handler, engine, graph)
    elif api_path == "/graph/data":
        _handle_graph_data(query, handler, engine, graph)
    else:
        send_error(handler, 404, "Route not found")


def handle_health(
    handler: BaseHTTPRequestHandler, engine: MemoryEngine | None,
    workspace: str = ""
) -> None:
    """GET /api/health — health check endpoint."""
    send_json(handler, {
        "status": "ok",
        "version": "0.2.0",
        "workspace": workspace,
        "memoryEnabled": engine is not None,
    })


def _handle_status(
    handler: BaseHTTPRequestHandler, engine: MemoryEngine | None
) -> None:
    """GET /api/memory/status."""
    if not engine:
        send_error(handler, 503, "Memory not initialized")
        return
    stats = engine.get_stats()
    tier_map: dict[str, int] = {}
    for t in stats["tier_breakdown"]:
        tier_map[t["tier"]] = t["entry_count"]
    send_json(handler, {
        "totalEntries": stats["total_entries"],
        "totalEdges": stats["total_edges"],
        "totalVectors": stats["total_vectors"],
        "tierBreakdown": tier_map,
    })


def _handle_search(
    query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
) -> None:
    """GET /api/memory/search?q=X."""
    if not engine:
        send_error(handler, 503, "Memory not initialized")
        return
    q = first_param(query, "q", "")
    limit = int(first_param(query, "limit", "10"))
    tier = first_param(query, "tier", "")
    results = (
        engine.search.search_in_tier(q, tier, limit)
        if tier
        else engine.search.search(q, limit)
    )
    send_json(handler, [to_entry_response(r["entry"], r["score"]) for r in results])


def _handle_list_entries(
    query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
) -> None:
    """GET /api/memory/entries?tier=X&type=Y&limit=Z&offset=W&sort=S."""
    if not engine:
        send_error(handler, 503, "Memory not initialized")
        return
    tier = first_param(query, "tier", "")
    type_ = first_param(query, "type", "")
    limit = int(first_param(query, "limit", "20"))
    offset = int(first_param(query, "offset", "0"))
    sort = first_param(query, "sort", "created_at")
    after_id = first_param(query, "after_id", "")
    entries = _query_entries(engine, tier, type_, limit, offset, sort, after_id)
    send_json(handler, [to_entry_response(e) for e in entries])


def _query_entries(
    engine: MemoryEngine, tier: str, type_: str,
    limit: int, offset: int, sort: str, after_id: str,
) -> list[dict]:
    """Build and execute filtered entry query."""
    sort_col = {"access_count": "access_count", "confidence": "confidence"}.get(
        sort, "created_at"
    )
    clauses = ["1=1"]
    params: list = []
    if tier:
        clauses.append("tier = ?")
        params.append(tier)
    if type_:
        clauses.append("type = ?")
        params.append(type_)
    if after_id:
        clauses.append("id > ?")
        params.append(int(after_id))
    where = " AND ".join(clauses)
    sql = f"SELECT * FROM knowledge_entries WHERE {where} ORDER BY {sort_col} DESC LIMIT ? OFFSET ?"
    params.extend([limit, offset])
    cur = engine._conn.execute(sql, params)
    return [dict(r) for r in cur.fetchall()]


def _handle_get_entry(
    api_path: str, handler: BaseHTTPRequestHandler, engine: MemoryEngine | None
) -> None:
    """GET /api/memory/entries/<id>."""
    if not engine:
        send_error(handler, 503, "Memory not initialized")
        return
    entry_id = int(api_path.split("/")[2])
    entry = engine.knowledge.find_by_id(entry_id)
    if not entry:
        send_error(handler, 404, "Not found")
        return
    engine.knowledge.record_access(entry_id)
    send_json(handler, to_detail_response(entry))


def _handle_neighbors(
    api_path: str, handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None, graph: KnowledgeGraph | None,
) -> None:
    """GET /api/memory/graph/<id>/neighbors."""
    if not engine or not graph:
        send_error(handler, 503, "Not initialized")
        return
    entry_id = int(api_path.split("/")[2])
    neighbor_ids = graph.get_connected(entry_id)
    entries = [engine.knowledge.find_by_id(nid) for nid in neighbor_ids]
    send_json(handler, [to_entry_response(e) for e in entries if e])


def _handle_graph_data(
    query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
    graph: KnowledgeGraph | None,
) -> None:
    """GET /api/memory/graph/data?limit=X — nodes from edges."""
    if not engine or not graph:
        send_error(handler, 503, "Not initialized")
        return
    limit = int(first_param(query, "limit", "500"))
    edges = engine.graph_repo.find_all(limit)
    if not edges:
        entries = engine.knowledge.find_by_tier("WORKING", 50)
        nodes = [to_graph_node(e) for e in entries]
        send_json(handler, {"nodes": nodes, "edges": []})
        return
    node_ids: set[int] = set()
    for e in edges:
        node_ids.add(e["source_id"])
        node_ids.add(e["target_id"])
    nodes = []
    for nid in node_ids:
        entry = engine.knowledge.find_by_id(nid)
        if entry:
            nodes.append(to_graph_node(entry))
    edge_list = [
        {"source": e["source_id"], "target": e["target_id"], "relation": e["relation"]}
        for e in edges
    ]
    send_json(handler, {"nodes": nodes, "edges": edge_list})


def _handle_audit(
    query: dict[str, list[str]],
    handler: BaseHTTPRequestHandler,
    engine: MemoryEngine | None,
) -> None:
    """GET /api/memory/audit?limit=20&after_id=X — recent audit events."""
    if not engine:
        send_error(handler, 503, "Memory not initialized")
        return
    limit = int(first_param(query, "limit", "20"))
    after_id = first_param(query, "after_id", "")
    if after_id:
        cur = engine._conn.execute(
            "SELECT * FROM memory_audit WHERE id > ? ORDER BY id DESC LIMIT ?",
            (int(after_id), limit),
        )
    else:
        cur = engine._conn.execute(
            "SELECT * FROM memory_audit ORDER BY id DESC LIMIT ?", (limit,)
        )
    events = [dict(r) for r in cur.fetchall()]
    send_json(handler, [
        {"id": e["id"], "operation": e["operation"], "entryId": e.get("entry_id"),
         "sessionId": e.get("session_id"), "details": e.get("details"),
         "createdAt": e["created_at"]}
        for e in events
    ])
